### PR TITLE
Elm frontend

### PR DESCRIPTION
**Pull request by pepeiborra**
Here we go again. 

This is a not-so-straight port of `debug.js` to `debug.elm`. From there, `debug.js` is generated with
```
pepe:~/scratch/debug$ cd elm
pepe:~/scratch/debug/elm$ elm-make Main.elm --output=../html/debug.js
```

I believe that there are no functional changes other than a minor fix to the generator of the #function-source control. And no DOM manipulation anymore ! This port is highly declarative and, I think, much easier to modify and maintain. I didn't really like the Elm language, but I enjoyed the Elm libraries and the so called architecture. In fact, this rewrite was a lot of fun.

I would recommend merging this over the Purescript rewrite of the DOM frontend. I like Purescript better as a language and I know that it has Elm-like frameworks, but I have run out of time to continue experimenting, and I am rather happy with the Elm rewrite tbh.

TODO
- Integrated build system that generates `debug.js` if the elm toolchain is in the path, please help !
- Add a CI step that validates the Elm code

### CODE REVIEW

File: elm/Main.elm [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+                            case match of
+                                "" ->
+                                    []
+
+                                "where" ->
+                                    span [ class "hs-keyword" ] [ text "where" ]
+                                        :: loop (String.dropLeft (String.length match) src)
+
+                                _ ->
+                                    (case String.indexes match symbols of
```
Comment (pepeiborra) [inline elm/Main.elm]: No pattern guards

File: elm/Trace.elm [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+decodeCallDataDict : DebugTrace a -> Dict String Value -> Result String CallData
+decodeCallDataDict trace d =
+    Result.map4 CallData
+        (get "$depends" d |> Maybe.map (Decode.decodeValue (list int)) |> Maybe.withDefault (Result.Ok []))
+        (get "$parents" d |> Maybe.map (Decode.decodeValue (list int)) |> Maybe.withDefault (Result.Ok []))
+        (get "" d
+         |> Result.fromMaybe "Corrupt trace: call without function id"
+         |> Result.andThen (Decode.decodeValue int)
+         |> Result.andThen (flip Array.get trace.functions >> Result.fromMaybe "Corrupt trace: missing function id"))
```
Comment (pepeiborra) [inline elm/Trace.elm]: No do notation or monad comprehensions :(

File: elm/Trace.elm [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+
+type alias DebugTrace a =
+    { functions : Array DebugFunction -- ^ Flat list of all the functions traced
+    , variables : Array String -- ^ Flat list of all the variable values observed
+    , calls : Array a -- ^ Flat list of all the function calls traced
+    }
+
+
+traverseDebugTrace : (a -> Result String b) -> DebugTrace a -> Result String (DebugTrace b)
```
Comment (pepeiborra) [inline elm/Trace.elm]: No rank-2 types :(

Comment (pepeiborra) [reply elm/Trace.elm]: nor type classes :(

File: elm/Main.elm [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+        CallFilter x ->
+            ( { model
+                | callFilter =
+                    if x == "" then
+                        Nothing
+                    else
+                        Just (regex x)
+              }
+            , Cmd.none
+            )
```
Comment (pepeiborra) [inline elm/Main.elm]: The model-update-view pattern is pretty neat

File: elm/Main.elm [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+                        , viewVariables trace (Maybe.map (\x -> x.call) processedCall)
+                        ]
+                    ]
+                , tr [ id "function-depends-section" ]
+                    [ td []
+                        [ h2 [] [ text "Calls" ]
+                        , viewCallStack trace (Maybe.map (\x -> x.call) processedCall)
+                        ]
+                    ]
+                ]
```
Comment (pepeiborra) [inline elm/Main.elm]: The declarative approach to defining the view is priceless

File: elm/Main.elm [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+        |> List.map
+            (\arg ->
+                case Dict.get arg call.callLocals of
+                    Nothing ->
+                        Just ( arg, "_" )
+
+                    Just valueIx ->
+                        Array.get valueIx trace.variables |> Maybe.map (\x -> ( arg, x ))
+            )
+        |> Maybe.combine
```
Comment (pepeiborra) [inline elm/Main.elm]: No `foldMap` (no `Monoid`s either)

**Comment by ndmitchell** [comment]
I like it! Code looks readable, and I've confirmed I can generate it locally, so happy to say we are now using Elm for reals.

**Pull request merged**